hw/dv/dpi/usbdpi/usbdpi.h

```diff
@@ -8,30 +8,58 @@
 #define TOOL_VERILATOR 1
 #define TOOL_INCISIVE 0
 
+#include <assert.h>
 #include <limits.h>
+#include <stdint.h>
 #include <stdio.h>
+#include <stdlib.h>
+#if USBDPI_STANDALONE
+typedef uint32_t svBitVecVal;
+#else
 #include <svdpi.h>
+#endif
+
+#include "usb_monitor.h"
+#include "usb_transfer.h"
 
-// How many bits in our frame (1ms on real hardware)
-#define FRAME_INTERVAL 256 * 8
+// Use a local macro for assertions so that we can redefine their behaviour if
+// required
+#ifndef USBDPI_ASSERT
+#define USBDPI_ASSERT assert
```

Comment:
Yes, possibly. I've discovered - the hard way - that assertion failures and invalid memory accesses in the DPI model do not cleanly kill the DPI thread and process, and then stop the simulation in the manner one would expect.
I therefore found it necessary to make assertion failures more obvious by changing their behaviour. I still need to raise an issue with someone who knows the opentitantool/Vchip_sim_tb interaction to tidy this up. What happens at present is that the test runs until it times out, and leaves no indication of what happened.

Reply:
Yeah... the existing code quality is very poor. Assertions go away when we compile normally, so they shouldn't be used for real error handling, but the original has the mistake in many places.

For opentitantool, there is effectively no interaction with Vchip_sim_tb for USB tests, other than spawning the process and monitoring the UART. The DPI module's architecture is very much not ideal, as it embeds a test vector in the testbench (instead of taking on the role of a host agent and providing an interface for another entity to drive a sequence).

Reply:
With our current build system, the assertions remain and there is a textual report of an assertion failure, buried somewhere in the test log, but the main problem is that an assertion failure or an invalid memory access does not actually terminate Vchip_sim_tb in the expected fashion; rather opentitantool spins and Vchip_sim_tb remains present but somnolent, until such time as the bazel test times out.
USBDPI_ASSERT helped to some degree in discovering the above issue, but presently serves no purpose, so I've reverted all uses to the standard assert macro.

Reply:
Thanks! Perhaps opentitantool needs to monitor the subprocess's state and get some indication of what an unexpected termination is (and count it as a failure). A problem for another day, heh.